### PR TITLE
increase number of supported cases for direct function calls

### PR DESCRIPTION
**Pull request by mjkkirschner**
### Purpose

This PR begins to address https://jira.autodesk.com/browse/DYN-5321, but we'll keep this open as there are remaining known cases which are still failing. Unclear how to fix some of these at the moment and they require more discussion.

1. PR attempts to unify custom compile time will replicate logic with DFSTraversal which is already pretty good at determining compile time type information. This ends up simplifying the `WillReplicate` checks. Uses two compile passes to achieve this.
2. When determining if call will replicate also includes the selected `FEP` as a parameter so we can check what attributes each parameter has - these are useful when determining if a FEP will replicate - for example the `arbitraryRankArrayImport` attribute should make it so a parameter is not replicated over except with specific replication guides.
3. PR attempts to replace `ArrayTypeStatistics` logic in `EmitExprListNode` method with DFSTraversal - simplifies that code and uses two compile passes to achieve this.
4. introduces `StrictDirectFunctionCallValidation` on the `MSILRuntime` and CodeGenIL` objects, if these bools are set true exceptions will be thrown when direct function calls should made, but can't. If these values are false we'll fallback to replicated calls.

### Declarations

Check these if you believe they are true

- [x] The codebase is in a better state after this PR
- [x] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [ ] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files
- [ ] All tests pass using the self-service CI.
- [ ] Snapshot of UI changes, if any.
- [ ] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions) and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.
- [ ] This PR modifies some build requirements and the readme is updated

### Release Notes

(FILL ME IN) Brief description of the fix / enhancement. **Mandatory section** 

### Reviewers

(FILL ME IN) Reviewer 1  (If possible, assign the Reviewer for the PR)

(FILL ME IN, optional) Any additional notes to reviewers or testers.

### FYIs

(FILL ME IN, Optional) Names of anyone else you wish to be notified of

### CODE REVIEW

File: src/Engine/EmitMSIL/CodeGenIL.cs [hunk 1]
```diff
@@ -1612,46 +1588,77 @@ private bool DoesParamArgRankMatch(List<Type> parameterTypes, List<AssociativeNo
...
                 }
                 else return false;
                 argIndex++;
             }
             return true;
         }
 
-        private static int GetRank(Type type)
+        private static bool DoesParamArgRankMatchInner(ProtoCore.CLRFunctionEndPoint fep,Type p, Type t, int paramIndex)
         {
```
Comment (aparajit-pratap) [inline src/Engine/EmitMSIL/CodeGenIL.cs:1626]: `DoesParamArgRankMatchInner(ProtoCore.CLRFunctionEndPoint fep,Type paramType, Type argType, int paramIndex)`

Comment (aparajit-pratap) [reply src/Engine/EmitMSIL/CodeGenIL.cs:1626]: Can we rename this function as the `inner` in its name sounded like it was matching the arg and param types of each element of an array?

File: src/Engine/EmitMSIL/CodeGenIL.cs [hunk 2]
```diff
@@ -1612,46 +1588,77 @@ private bool DoesParamArgRankMatch(List<Type> parameterTypes, List<AssociativeNo
...
+
+        private static int GetRank(ProtoCore.CLRFunctionEndPoint fep,Type type,int paramIndex)
+        {
+            if(type == typeof(IList)){
+                return -1;
+            }
+            if(type == typeof(object) && fep.ParamAttributes[paramIndex].Any(attr => attr is ArbitraryDimensionArrayImportAttribute))
+            {
+                return -1;
+            }
```
Comment (aparajit-pratap) [inline src/Engine/EmitMSIL/CodeGenIL.cs:1661]: I believe a parameter of type `IEnumerable<T>`  with `ArbitraryDimensionImportAttribute` results in an arbitrary dimension array of type `T`, in which case the `type == typeof(object)` check doesn't seem quite right, does it?

Comment (mjkkirschner) [reply src/Engine/EmitMSIL/CodeGenIL.cs:1661]: Hi @aparajit-pratap, this check is specifically for a test that was calling `List.Contains` - this method and some of the other core list methods accept object parameters marked with this attribute:
https://github.com/DynamoDS/Dynamo/blob/c9a23ab9c9e4110ac798b33a9d4afad183718034/src/Libraries/CoreNodes/List.cs#L51

File: src/Engine/EmitMSIL/CodeGenIL.cs [hunk 3]
```diff
@@ -1585,11 +1555,17 @@ private void EmitUnmarshalFunctionArgs(List<AssociativeNode> args, List<Type> pa
...
+        private bool DoesParamArgRankMatch(ProtoCore.CLRFunctionEndPoint fep, List<Type> parameterTypes, List<AssociativeNode> args)
         {
             var argIndex = 0;
+          
             foreach (var p in parameterTypes)
             {
+                var currentArg = args[argIndex];
+                if (compilePass == CompilePass.GatherTypeInfo)
+                {
+                    DfsTraverse(currentArg);
```
[the remark this answers is not in the document]
Comment (mjkkirschner) [reply src/Engine/EmitMSIL/CodeGenIL.cs:1567]: yes, exactly.

File: src/Engine/EmitMSIL/CodeGenIL.cs (hunk 1 above)
Comment (mjkkirschner) [reply src/Engine/EmitMSIL/CodeGenIL.cs:1626]: yeah, I'll think of something better.

File: src/Engine/EmitMSIL/CodeGenIL.cs (hunk 2 above)
Comment (aparajit-pratap) [reply src/Engine/EmitMSIL/CodeGenIL.cs:1678]: I guess my question is how do we deal with parameters like this:
```
public void foo([ArbitraryDimensionArrayImport] IEnumerable<int> bar) { ... }
```
that are passed an `int` or list of any rank of `int`s as argument.

Comment (mjkkirschner) [reply src/Engine/EmitMSIL/CodeGenIL.cs:1678]: @aparajit-pratap I need to test it, but I agree it's a concern for example - `FirstItem(list:var[]..[]` should work with a list or a single item, I'm not sure what happens today though, I can add a test for that case at the minimum.

Comment (mjkkirschner) [reply src/Engine/EmitMSIL/CodeGenIL.cs:1665]: ah @aparajit-pratap to clarify, are you asserting that because `int` is a `valueType` that my check with `object` is actually too strict?

Now that I think about it - it does make sense that basically any use of this attribute on a parameter means it cannot replicate - EXCEPT in cases where there are replicate guides that force an iteration - does that sound right?

Comment (aparajit-pratap) [reply src/Engine/EmitMSIL/CodeGenIL.cs:1665]: 1. I believe there are 2 cases where ArbitraryDimension attribute applies, where it converts from:
a. `IList` to `var[]..[]` - I think you've taken care of this
b. `IEnumerable<T>` to `T[]..[]` - my question was about this case, how would you handle it as I don't think checking for `object` type would work here
2. I'm unsure what happens when it is applied to `object`. If it does, your check for the type being equal to `object` makes sense.
3. I agree, it would be good to test this logic with a single item arg passed to a parameter with the attribute

File: src/Engine/EmitMSIL/CodeGenIL.cs [hunk 4]
```diff
@@ -1634,46 +1639,105 @@ private bool DoesParamArgRankMatch(List<Type> parameterTypes, List<AssociativeNo
...
+                    return false;
+                }
+
+            }
+            return true;
+        }
+
+        private static int GetRank(ProtoCore.CLRFunctionEndPoint fep,Type type,int paramIndex)
         {
+            //TODO (Alternatives to consider)
```
Comment (mjkkirschner) [inline src/Engine/EmitMSIL/CodeGenIL.cs:1723]: I'd like some feedback on these alternatives
1. we use the imported protocore typesystem and just check if the imported rank is arbitrary
2. we try to minimally recreate the logic that the type system encodes to try to remove our dependency on it.

**Comment by mjkkirschner** [comment]
Even though as you can see from the comments this PR leaves many failures when `StrictDirectFunctionCallValidation` is enabled, I would like to get this PR in as I believe it's an improvement on several fronts in code simplification and does fix a bunch of cases.

I can file another followup for the remaining classes of problems - I think they are bigger problems that will require discussion - in addition some are already in progress and will be resolved by compile time overload resolution.

File: src/Engine/EmitMSIL/CodeGenIL.cs [hunk 5]
```diff
@@ -1380,21 +1358,48 @@ private void EmitUnmarshalFunctionArgs(List<AssociativeNode> args, List<Type> pa
...
+        /// Checks rank, type, and attributes to make this determination.
+        /// </summary>
+        /// <param name="fep"></param>
+        /// <param name="parameterTypes"></param>
+        /// <param name="args"></param>
+        /// <returns></returns>
+        /// <exception cref="Exception"></exception>
+        private bool DoParamsRankMatchArgs(ProtoCore.CLRFunctionEndPoint fep, List<Type> parameterTypes, List<AssociativeNode> args)
         {
+            //TODO this can throw because it's possible that the FEP is incorrect
```
[the remark this answers is not in the document]
Comment (mjkkirschner) [reply src/Engine/EmitMSIL/CodeGenIL.cs:1372]: yes

File: src/Engine/EmitMSIL/CodeGenIL.cs [hunk 6]
```diff
@@ -1408,46 +1413,105 @@ private bool DoesParamArgRankMatch(List<Type> parameterTypes, List<AssociativeNo
...
+                }
+
+            }
+            return true;
+        }
+
+        private static int GetRank(ProtoCore.CLRFunctionEndPoint fep,Type type,int paramIndex)
         {
+            //TODO (Alternatives to consider)
+            if (fep.procedureNode.ArgumentTypes.ElementAt(paramIndex).rank == DSASM.Constants.kArbitraryRank)
```
[the remark this answers is not in the document]
Comment (mjkkirschner) [reply src/Engine/EmitMSIL/CodeGenIL.cs:1498]: while yes, that is true, IMO in the case of an arbitrary rank parameter, it does not matter what the arg rank is, we can mark them as compatible as we should not replicate.

If you prefer the alternative check that uses the individual arg/param types then I can switch to that, though I expect it will eventually get more complex or perhaps I can find a way of converting the clr type to a dstype and querying whether it is arbitrary rank.

File: src/Engine/EmitMSIL/CodeGenIL.cs [hunk 7]
```diff
@@ -1408,46 +1413,105 @@ private bool DoesParamArgRankMatch(List<Type> parameterTypes, List<AssociativeNo
...
+                {
+                    return true;
+                }
+
+                var argRank = GetRank(fep,argType,paramIndex);
+                var paramRank = GetRank(fep,paramType, paramIndex);
+
+                //if the param is an arbitrary rank array and
+                //arg is some array type, replication should not occur.
+                if (paramRank == -1 && argRank > 0)
```
[the remark this answers is not in the document]
Comment (mjkkirschner) [reply src/Engine/EmitMSIL/CodeGenIL.cs:1475]: yes, rank 0 would just be a single item I believe.

One thing that is unclear to me -

If a parameter of rank -1 (arbitrary) will never replicate with a single item OR an array, then this check is unnecessary.

@aparajit-pratap is that always the case?

[the remark this answers is not in the document]
Comment (aparajit-pratap) [reply src/Engine/EmitMSIL/CodeGenIL.cs:1475]: Comparing a single parameter to an argument, if the parameter has the arbitrary rank attribute, then it doesn't matter what rank the argument has, it will not replicate. However, for a function with more than one parameter, there can be one parameter with the attribute, and another one that doesn't. In this case, the function can still replicate over the parameter that doesn't have the attribute. So only if `DoesParamRankMatchArg` returns true for all parameter-argument pairs for a function call, then we can decide not to replicate.

File: src/Engine/EmitMSIL/CodeGenIL.cs (hunk 2 above)
Comment (aparajit-pratap) [reply src/Engine/EmitMSIL/CodeGenIL.cs:1507]: It looks like the `ArbitraryDimension` attribute applies to `object` as well, so you're right about checking for object type. Scratch (2) above.

File: src/Engine/EmitMSIL/CodeGenIL.cs [hunk 8]
```diff
@@ -1380,21 +1358,48 @@ private void EmitUnmarshalFunctionArgs(List<AssociativeNode> args, List<Type> pa
...
+            
             foreach (var p in parameterTypes)
             {
-                if (CoreUtils.IsPrimitiveASTNode(args[argIndex]) || args[argIndex] is CharNode)
+                var currentArg = args[argIndex];
+                if (compilePass == CompilePass.GatherTypeInfo)
+                {
+                    DfsTraverse(currentArg);
+                }
+                if (CoreUtils.IsPrimitiveASTNode(args[argIndex]) || currentArg is CharNode)
```
Comment (pinzart90) [inline src/Engine/EmitMSIL/CodeGenIL.cs:1394]: maybe we can use `currentArg` instead of `args[argIndex]` here

File: src/Engine/EmitMSIL/CodeGenIL.cs (hunk 7 above)
Comment (pinzart90) [inline src/Engine/EmitMSIL/CodeGenIL.cs:1475]: can we use Constants.kArbitraryRank instead of -1 ?

File: src/Engine/EmitMSIL/CodeGenIL.cs (hunk 4 above)
Comment (pinzart90) [reply src/Engine/EmitMSIL/CodeGenIL.cs:1497]: Well we did have discussions about removing the dependency on the old VM.
Not sure what that means now, maybe we can share some of that logic or re-write it.
I think we can use the protocore system now and we can separate out later

File: src/Engine/EmitMSIL/CodeGenIL.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (mjkkirschner) [reply src/Engine/EmitMSIL/CodeGenIL.cs:1475]: ahh, good call, let me see if I can write a test for that case and address it.

**Comment by mjkkirschner** [comment]
closing this in favor of emitMSIL-v2 version. Will try to address review comments there.
